### PR TITLE
Fix distributed store to use add for the counter of DL shared seed

### PR DESCRIPTION
**Pull request by ejguan**
In order to get the result of `_shared_seed_recv_cnt` properly, switch from `store.get` to `store.add(key, 0)`.

See the comment from distributed team for the reason:
https://github.com/pytorch/pytorch/blob/590d3e5774110e4657dcaa6acdb387ef69e41b47/torch/distributed/distributed_c10d.py#L242-L246

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -567,20 +572,26 @@ def _get_shared_seed(self):
                 ws = dist.get_world_size()
                 store = dist.distributed_c10d._get_default_store()
                 if rank == 0:
-                    store.set("_dl_shared_seed", str(_shared_seed))
+                    _shared_seed_str = str(_shared_seed)
+                    store.set("_dl_shared_seed", _shared_seed_str)
```
Comment (VitalyFedyunin) [inline torch/utils/data/dataloader.py:576]: super nit: keys might be module level constants

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -567,20 +572,26 @@ def _get_shared_seed(self):
...
+                    logger.info(f"Shared seed ({_shared_seed_str}) sent to store on rank 0")
+                    # Use 'add' instead of 'get' since for some store implementations 'add'
+                    # doesn't work well with 'get'.
+                    _shared_seed_recv_cnt = store.add("_dl_shared_seed_recv_cnt", 1)
+                    while _shared_seed_recv_cnt < ws:
+                        time.sleep(0.01)
```
Comment (VitalyFedyunin) [inline torch/utils/data/dataloader.py:582]: snit: Sleep interval is good candidate to be constant

**Review by VitalyFedyunin — APPROVED** [review]

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:576]: Sounds great. Will change it.

File: torch/utils/data/dataloader.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:582]: Makes sense.

**Review by NivekT — APPROVED** [review]

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -567,20 +572,26 @@ def _get_shared_seed(self):
...
                     store.set("_dl_shared_seed", "")
-                    store.add("_dl_shared_seed_recv_cnt", -ws)
-                    assert int(store.get("_dl_shared_seed_recv_cnt")) == 0
+                    _shared_seed_recv_cnt = store.add("_dl_shared_seed_recv_cnt", -ws)
+                    assert _shared_seed_recv_cnt == 0
                 else:
                     _shared_seed_str = ""
                     store.wait(["_dl_shared_seed"], _utils.MP_STATUS_CHECK_INTERVAL)
                     while len(_shared_seed_str) == 0:
+                        time.sleep(0.01)
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/utils/data/dataloader.py]: Learned by the implementation from Distributed team. https://github.com/pytorch/pytorch/blob/590d3e5774110e4657dcaa6acdb387ef69e41b47/torch/distributed/distributed_c10d.py#L251
